godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,480 @@
+        p_name: GDExtensionConstStringNamePtr,
+        p_value: GDExtensionConstVariantPtr,
+    ) -> GDExtensionBool {
+        let instance = p_instance as *mut ScriptInstanceData<T>;
+        let name = StringName::from_arg_ptr(p_name as GDExtensionTypePtr, PtrcallType::Standard);
+        let value = &*Variant::ptr_from_sys(force_mut_ptr(p_value));
+
+        let result = (*instance).inner.borrow_mut().set(name.clone(), value) as u8;
+
+        result
+    }
+
+    pub(super) unsafe extern "C" fn get_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        p_name: GDExtensionConstStringNamePtr,
+        r_ret: GDExtensionVariantPtr,
+    ) -> GDExtensionBool {
+        let instance = p_instance as *const ScriptInstanceData<T>;
+        let name = StringName::from_arg_ptr(p_name as GDExtensionTypePtr, PtrcallType::Standard);
+
+        let result = (*instance)
+            .inner
+            .borrow()
+            .get(name.clone())
+            .map(|variant| {
+                variant.move_return_ptr(
+                    r_ret as GDExtensionTypePtr,
+                    godot_ffi::PtrcallType::Standard,
+                )
+            })
+            .is_some();
+
+        result as u8
+    }
+
+    pub(super) unsafe extern "C" fn get_property_list_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        r_count: *mut u32,
+    ) -> *const GDExtensionPropertyInfo {
+        let instance = p_instance as *const ScriptInstanceData<T>;
+
+        let property_list: Vec<_> = (*instance)
+            .inner
+            .borrow()
+            .property_list()
+            .iter()
+            .map(|prop| prop.property_sys())
+            .collect();
+
+        *r_count = property_list.len() as u32;
+
+        let ptr = Box::into_raw(property_list.into_boxed_slice());
+
+        (*ptr).as_ptr()
+    }
+
+    pub(super) unsafe extern "C" fn get_method_list_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        r_count: *mut u32,
+    ) -> *const GDExtensionMethodInfo {
+        let instance = p_instance as *const ScriptInstanceData<T>;
+
+        let method_list: Vec<_> = (*instance)
+            .inner
+            .borrow()
+            .method_list()
+            .iter()
+            .map(|method| method.method_sys())
+            .collect();
+
+        *r_count = method_list.len() as u32;
+
+        let ptr = Box::into_raw(method_list.into_boxed_slice());
+
+        (*ptr).as_ptr()
+    }
+
+    pub(super) unsafe extern "C" fn free_property_list_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        p_prop_info: *const GDExtensionPropertyInfo,
+    ) {
+        let instance = p_instance as *const ScriptInstanceData<T>;
+        let length = (*instance).inner.borrow().property_list().len();
+
+        let vec = Vec::from_raw_parts(p_prop_info as *mut GDExtensionPropertyInfo, length, length);
```

Comment:
this doesn't seem safe to me, assuming this `Vec` is the one that was allocated in `get_property_list_func`, then you don't set the capacity to be the same as length anywhere. so there is no guarantee that the length is the same as the capacity here.

Reply:
`into_boxed_slice()` in `get_property_list_func` ensures capacity and length are equal.

Reply:
Please document invariants that the caller must upheld with
```rs
/// # Safety
/// ...
```
in the function documentation, and on call site _how_ they are upheld using
```rs
// SAFETY
unsafe { my_func(...) };
```

Here there's no call site, so you can also just add a comment to the implementation itself (basically what you just wrote, above `Vec::from_raw_parts`).